src/rebuild/scan.c

```diff
@@ -384,9 +384,57 @@ rebuild_obj_scan_cb(daos_handle_t ch, vos_iter_entry_t *ent,
 					       rpt->rt_rebuild_ver,
 					       tgts, shards,
 					       rpt->rt_tgts_num, myrank);
+	} else if (rpt->rt_rebuild_op == RB_OP_RECLAIM) {
+		struct pl_obj_layout *layout = NULL;
+		bool still_needed;
+
+		/*
+		 * Compute placement for the object, then check if the layout
+		 * still includes the current rank. If not, the object can be
+		 * deleted/reclaimed because it is no longer reachable
+		 */
+		rc = pl_obj_place(map, &md, NULL, &layout);
+		if (rc != 0)
+			D_GOTO(out, rc);
+
+		still_needed = pl_obj_layout_contains(rpt->rt_pool->sp_map,
+						      layout, myrank);
+		if (!still_needed) {
+			D_DEBUG(DB_REBUILD, "deleting object "DF_UOID
+				" which is no longer reachable on this rank",
+				DP_UOID(oid));
+			/*
+			 * It's possible this object might still be being
+			 * accessed elsewhere - retry until until it is possible
+			 * to delete
+			 */
+			do {
+				rc = vos_obj_delete(param->ip_hdl, oid);
```

Comment:
hmm, you may need reprobe the tree if you delete the object from oi tree.  see migrate_cont_iter_cb().

Reply:
Interesting, I'll take a look at this.

Reply:
Right, see vos_iterate(), if iteration callback deletes entry or yield, it needs to inform vos_iterate() by setting VOS_ITER_CB_DELETE or VOS_ITER_CB_YIELD, so that vos_iterate() can determing if it's safe to move next or required a re-probe.

Reply:
Fixed